service-daemon

```diff
@@ -1,47 +1,97 @@
-#!/data/data/com.termux/files/usr/bin/sh
+#!/usr/bin/env sh
+# shellcheck shell=dash
 
-set -e
-start(){
-    mkdir -p $PIDDIR
-    start-stop-daemon -S -b -m -p $PIDFILE -x $DAEMON -d $PREFIX -- $DAEMON_OPTS
+set -eu
+
+_prefix="${TERMUX__PREFIX:-"${PREFIX}"}"
+
+: "${XDG_RUNTIME_DIR:="${_prefix}/var/run"}"
+export XDG_RUNTIME_DIR
+
+BINDIR="${_prefix}/bin"
+readonly BINDIR
+
+LOGDIR="${_prefix}/var/log"
+export LOGDIR
+readonly LOGDIR
+
+SVDIR="${_prefix}/var/service"
+export SVDIR
+readonly SVDIR
+
+unset -v _prefix
+
+
+wait_for_pid() {
+    local pid="$1"
+    local seconds
+
+    for seconds in 1 2 3 4 5 ; do
+        kill -0 "${pid}" 2>/dev/null || break
+        sleep "${seconds}"
+    done
+    kill -0 "${pid}" 2>/dev/null && return 1 || return 0
 }
-stop(){
-    start-stop-daemon -K -s 1 -o -p $PIDFILE
-    tail -f --pid=`cat $PIDFILE` /dev/null
-    for f in $SVDIR/*/supervise $SVDIR/*/log/supervise;do
-        kill -9 `cat $f/pid` 2>/dev/null ||:
-        rm -rf $f
+
+kill_from_pid_file() {
+    local pid
+
+    for pid in $(head -n 1 "$1") ; do
+        kill "${pid}"
+        wait_for_pid "${pid}" || kill -9 "${pid}"
     done
 }
+
+start() {
+    # shellcheck disable=SC2174
+    mkdir -m 0700 -p "${XDG_RUNTIME_DIR}"
+    start-stop-daemon -S -b -m -p "${PIDFILE}" -x "${DAEMON}" -d "${XDG_RUNTIME_DIR}" -- "${DAEMON_OPTS}"
+}
+
+stop() {
+    local dir pid="$(cat "${PIDFILE}")"
+
+    start-stop-daemon -K -s "${DAEMON_STOP_SIGNAL}" -o -p "${PIDFILE}"
+    wait_for_pid "${pid}" || tail -f --pid="${pid}" /dev/null
+    for dir in "${SVDIR}"/*/{,log/}supervise ; do
+        kill_from_pid_file "${dir}/pid"
+        rm -rf "${dir}"
+    done
+}
+
+
 # Must be a valid filename
-NAME=service-daemon
-PIDDIR=$PREFIX/var/run
-PIDFILE=$PIDDIR/$NAME.pid
-#This is the command to be run, give the full pathname
-DAEMON=$PREFIX/bin/runsvdir
-DAEMON_OPTS="$SVDIR"
+NAME='service-daemon'
+PIDFILE="${XDG_RUNTIME_DIR}/${NAME}.pid"
+# This is the command to be run, give the full pathname
+DAEMON="${BINDIR}/runsvdir"
+DAEMON_OPTS="${SVDIR}"
+# If runsvdir receives a TERM signal, it exits with 0 immediately.
+# If runsvdir receives a HUP signal,
+# it sends a TERM signal to each runsv(8) process it is monitoring and then exits with 111.
+DAEMON_STOP_SIGNAL='HUP'
+
 
 case "$1" in
-    start)
-        echo -n "Starting daemon: "$NAME
-        start
-        echo "."
-        ;;
-    stop)
-        echo -n "Stopping daemon: "$NAME
-        stop
-        echo "."
+    (restart) printf -- 'Restart' ;;
+    (start) printf -- 'Start' ;;
+    (stop) printf -- 'Stopp' ;;
```

Comment:
This is still harming the readability of this code.
I do not see a benefit is keeping this printout DRY to such an extent.

I'm also kind of surprised to see that the enclosed parentheses form for case labels is part of the POSIX sh spec.
But I did check it in the documentation.

Reply:
I think the new function is good balance between duplicating in every case and the extent the previous version went to.

Let me know if you have a better way to maintain readability without duplicated messages for every case.

The open parentheses always look odd to me, so when I learned about the closed form, I was happy to start using that way instead.